### ISSUE TITLE
cwrap generated functions allocate string args on wasm stack, not good for huge strings

### ISSUE THREAD

**Issue by pmachek** [emscripten-core/emscripten#20015]
Please include the following in your bug report:

**Version of emscripten/emsdk:**
emcc (Emscripten gcc/clang-like replacement + linker emulating GNU ld) 3.1.43 (a6b8143cf3c1171db911750359456b15a8deece7)
clang version 17.0.0 (https://github.com/llvm/llvm-project 71513a71cdf380efd6a44be6939e2cb979a62407)
Target: wasm32-unknown-emscripten
Thread model: posix

**Failing command line in full:**
emcc <my cpp files> -O3 -o example.js -s EXPORTED_FUNCTIONS=_myFunction,_myFunction2,_myFunction3,_malloc,_free -s ASSERTIONS=1 -s EXPORTED_RUNTIME_METHODS=ccall,cwrap,stringToUTF8 -s ALLOW_MEMORY_GROWTH=1 -pedantic

**Full link command and output with `-v` appended:**
Probably not relevant.

The problem:
I can wrap myFunction via cwrap
```
const myFunction = Module.cwrap(
    "myFunction",
    "number",
    ["string"]
);
```

If I attempt to call myFunction with a string argument that is ~1.45MB in size, I get RuntimeError: index out of bounds raised from createExportWrapper. I remember that at some point of my experimenting I was able to see a crash after a function call suggesting that allocation of string was done on stack. So I tried to change the code to allocation on heap
```
      const size = lengthBytesUTF8(str) + 1;
      const ptr = Module._malloc(size);
      Module.stringToUTF8(str, ptr, size);
      Module._myFunction(ptr);
      Module._free(ptr);
```
And with this change everything works as expected. Unfortunately, I don't have the exact stack trace, so it is still possible that I misread the situation, but the fact that everything works now aligns with the assumption that the problem comes from overflowing the stack.

**Comment by pmachek** [comment]
I think I got the stack trace suggesting stack allocation of string when I experimented with the exact size of string I can pass before the code crashes.

**Comment by kripken** [comment]
I believe we do allocate on the stack there. In a build with assertions (or perhaps only with `STACK_OVERFLOW_CHECK=2`) you should get a runtime error if you allocate too much.

Ensuring a large-enough stack should avoid this, and would still be more efficient than heap allocation for this kind of thing, I think. But I agree this is not a perfect situation.

**Comment by pmachek** [comment]
So, how about having a limit for size you allocate on stack? You'd still allocate on the stack most of the time, keeping the hot path as-is. Or better, if checking the stack capacity is reasonably cheap operation, the code could make sure that at least some predetermined capacity will be left on the stack and if not, it would switch to allocation on heap.

**Comment by pmachek** [comment]
Anyway,
a) It is unexpected that cwrap produces errors where safe handling of situation is easily possible. Avoiding the problem will cost tiny bit of performance, but that is to be expected of the convenience function. People who want to save every possible nanosecond should not rely on convenience wrappers. But more casual users do expect some basic safety from the wrapper.
b) It is unexpected that I get RuntimeError: index out of bounds, rather then more refined exception (StackOverflow: Attempted to access memory beyond the stack limits, something like that...?)

**Comment by sbc100** [comment]
In debug builds (e.g. when building with either `-O0` or explicitly with `-sASSERTIONS`) you should see a clear error message which includes information on how to increase the limit.  Do you not see this message in you debug build?: https://github.com/emscripten-core/emscripten/blob/ef3fd077475a5124cf3dd8047f1f34bcaa05e333/src/library.js#L3045-L3047

**Comment by kripken** [comment]
@pmachek 

> So, how about having a limit for size you allocate on stack? 

That's a good idea, but a downside is that it would mean including `malloc/free` in all builds that use `ccall/cwrap`, which in small programs can be a significant amount of code size, unfortunately.

Overall there are tricky tradeoffs here, but I still think it's best to have errors in debug builds, as I think we have as @sbc100 said - if those aren't working or aren't clear enough we should definitely fix those!

**Comment by sbc100** [comment]
> @pmachek
> 
> > So, how about having a limit for size you allocate on stack?
> 
> That's a good idea, but a downside is that it would mean including `malloc/free` in all builds that use `ccall/cwrap`, which in small programs can be a significant amount of code size, unfortunately.

Yes, I think I would be against adding this complexity and code size to cwrap.  As long as we give good error messages I think that should be enough.

> 
> Overall there are tricky tradeoffs here, but I still think it's best to have errors in debug builds, as I think we have as @sbc100 said - if those aren't working or aren't clear enough we should definitely fix those!

**Comment by pmachek** [comment]
@sbc100 I see, interesting. I'm running this code as a Firefox extension and there's nothing like the abort message in the browser console. All I get is the RuntimeError: index out of bounds. As you can see, I'm compiling with assertions, but I did not set optimizations to -O0

@kripken Makes sense in the context. If I got the error message about stack overflow explaining how to set stack myself, I would be just fine.

**Comment by sbc100** [comment]
It would be interesting to see why you didn't see the error message.

Can you trying building with `-O0` ?  

Do you see the `If you require more stack space build with -sSTACK_SIZE=<bytes>` message in the generated JS file?
